test/e2e/shoot/logging/common.go

```diff
@@ -0,0 +1,166 @@
+// If wantedEventMessages is empty than the function will return nill and will not query the Loki for any events.
+func WaitUntilLokiReceivesEvent(ctx context.Context, logger logr.Logger, interval time.Duration, lokiLabels map[string]string, tenant, lokiNamespace, queryAfterUnpack string, wantedEventMessages []string, startTime time.Time, c kubernetes.Interface) error {
+	if len(wantedEventMessages) < 1 {
+		return nil
+	}
+
+	query := `{job="event-logging"} | unpack`
+	if queryAfterUnpack != "" {
+		query = query + ` | ` + queryAfterUnpack
+	}
+	log := logger.WithValues("query", query)
+
+	return retry.Until(ctx, interval, func(ctx context.Context) (done bool, err error) {
+		search, err := GetLokiLogsWithCMD(ctx, log, lokiLabels, tenant, lokiNamespace, query, startTime, c)
+		if err != nil {
+			return retry.SevereError(err)
+		}
+
+		if len(search.Data.Result) < len(wantedEventMessages) {
+			log.Info("Waiting to receive all expected logs", "events", len(search.Data.Result), "wanted", len(wantedEventMessages))
+			return retry.MinorError(fmt.Errorf("received only %d/%d events", len(search.Data.Result), len(wantedEventMessages)))
+		} else if len(search.Data.Result) > len(wantedEventMessages) {
+			return retry.SevereError(fmt.Errorf("expected to receive %d events but was %d", len(wantedEventMessages), len(search.Data.Result)))
+		}
+
+		var aggErr errorsutil.AggregateError
+		for _, wantedEventMessage := range wantedEventMessages {
+			found := false
+			for _, result := range search.Data.Result {
+				currentMessages := getAllStringsFromRangeSearchResponse(result.Values)
+				for _, currentMessage := range currentMessages {
+					if currentMessage == wantedEventMessage {
+						found = true
+						break
+					}
+				}
+				if found {
+					break
+				}
+			}
+			if !found {
+				aggErr = append(aggErr, fmt.Errorf("event with message \"%s\" is missing", wantedEventMessage))
+			}
+		}
+
+		if aggErr != nil {
+			return retry.SevereError(aggErr)
+		}
+
+		return retry.Ok()
+	})
+}
+
+func getAllStringsFromRangeSearchResponse(values [][]interface{}) []string {
+	var result []string
+	for _, interfaceEntry := range values {
+		for _, subInterfaceEntry := range interfaceEntry {
+			entry, ok := subInterfaceEntry.(string)
+			if !ok {
+				continue
+			}
+			result = append(result, entry)
+		}
+	}
+	return result
+}
+
+// GetLokiLogsWithCMD gets logs from startTime until now for query to loki instance in <lokiNamespace>
+func GetLokiLogsWithCMD(ctx context.Context, logger logr.Logger, lokiLabels map[string]string, tenant, lokiNamespace, query string, startTime time.Time, client kubernetes.Interface) (*RangeSearchResponse, error) {
+	lokiLabelsSelector := labels.SelectorFromSet(labels.Set(lokiLabels))
+
+	if tenant == "" {
+		tenant = "fake"
+	}
+
+	command := fmt.Sprintf("wget 'http://localhost:%d/loki/api/v1/query_range?start=%d' -O- '--header=X-Scope-OrgID: %s' --post-data='query=%s'", lokiPort, startTime.UnixNano(), tenant, query)
```

Comment:
Can we talk to loki directly via its `Ingress` instead of doing this hacky `wget` exec in the pod (which is not as stable)

Reply:
Loki does not have ingress. The ingress which it has is only for pushing logs from the shoot node so in this case, Loki is not accessible outside the seed.